src/resolver/data_url.zig

```diff
@@ -127,6 +128,114 @@ pub const DataURL = struct {
             return buf;
         }
 
-        return allocator.dupe(u8, percent_decoded);
+        return try allocator.dupe(u8, percent_decoded);
+    }
+
+    /// Returns the shorter of either a base64-encoded or percent-escaped data URL
+    pub fn encodeStringAsShortestDataURL(allocator: Allocator, mime_type: []const u8, text: []const u8) []u8 {
+        // Calculate base64 version
+        const base64_encode_len = bun.base64.encodeLen(text);
+        const total_base64_encode_len = "data:".len + mime_type.len + ";base64,".len + base64_encode_len;
+
+        use_base64: {
+            var counter = CountingBuf{};
+            const success = encodeStringAsPercentEscapedDataURL(&counter, mime_type, text) catch unreachable;
+            if (!success) {
+                break :use_base64;
+            }
+
+            if (counter.len > total_base64_encode_len) {
+                break :use_base64;
+            }
+
+            var buf = std.ArrayList(u8).init(allocator);
+            errdefer buf.deinit();
+            const success2 = encodeStringAsPercentEscapedDataURL(&buf, mime_type, text) catch unreachable;
+            if (!success2) {
+                break :use_base64;
+            }
+            return buf.items;
+        }
+
+        const base64buf = allocator.alloc(u8, total_base64_encode_len) catch bun.outOfMemory();
+        return std.fmt.bufPrint(base64buf, "data:{s};base64,{s}", .{ mime_type, text }) catch unreachable;
+    }
+
+    const CountingBuf = struct {
+        len: usize = 0,
+
+        pub fn appendSlice(self: *CountingBuf, slice: []const u8) Allocator.Error!void {
+            self.len += slice.len;
+        }
+
+        pub fn append(self: *CountingBuf, _: u8) Allocator.Error!void {
+            self.len += 1;
+        }
+
+        pub fn toOwnedSlice(_: *CountingBuf) Allocator.Error![]u8 {
+            return "";
+        }
+    };
+
+    pub fn encodeStringAsPercentEscapedDataURL(buf: anytype, mime_type: []const u8, text: []const u8) !bool {
+        const hex = "0123456789ABCDEF";
+
+        try buf.appendSlice("data:");
+        try buf.appendSlice(mime_type);
+        try buf.append(',');
+
+        // Scan for trailing characters that need to be escaped
+        var trailing_start = text.len;
+        while (trailing_start > 0) {
+            const c = text[trailing_start - 1];
+            if (c > 0x20 or c == '\t' or c == '\n' or c == '\r') {
+                break;
+            }
+            trailing_start -= 1;
+        }
+
+        var i: usize = 0;
+        var run_start: usize = 0;
+
+        while (i < text.len) {
```

Comment:
actually delete this entirely

if we need to validate unicode, use simdutf

bun.simdutf.validate.utf8